### PR TITLE
Add gru cudnn

### PR DESCRIPTION
**Pull request by zhaoyuchen2018**
1. Add cudnn gru op
  2. Add mul fc cudnn fusion, disable it for now
     as model use different gru algorithm.
  3. Need refine cudnn gru with lod info.
  test=develop

### CODE REVIEW

File: paddle/fluid/inference/api/paddle_pass_builder.h [hunk 1]
```diff
@@ -141,6 +142,9 @@ class GpuPassStrategy : public PassStrategy {
         "conv_elementwise_add_act_fuse_pass",        //
         "conv_elementwise_add2_act_fuse_pass",       //
         "conv_elementwise_add_fuse_pass",            //
+        // Disable cudnn gru fuse as model use different gru
```
Comment (Superjomn) [inline paddle/fluid/inference/api/paddle_pass_builder.h]: so these passes can't be used in the normally trained model?

Why not remove these three lines and call InsertPass to activate  them only in your test?

File: paddle/fluid/operators/cudnn_gru_op.cu.cc [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    LOG(INFO) << "stderr, CUDA Error: " << cudaGetErrorString(stat)
+              << " file:" << file << " line:" << line;
+  }
+}
+
+template <typename T>
+class CudnnGRUGPUKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext &ctx) const override {
+    auto *x = ctx.Input<LoDTensor>("Input");
```
Comment (Superjomn) [inline paddle/fluid/operators/cudnn_gru_op.cu.cc]: Too long function implementation, please split to several small ones to make the codes more clear to understand.

File: paddle/fluid/framework/ir/fc_cudnn_gru_fuse_pass.cc [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+    // copy update gate data
+    memcpy(w_data, weight_x_tensor->data<float>(), offset * sizeof(float));
+    w_data += offset;
+
+    memcpy(w_data, weight_x_tensor->data<float>() + 2 * offset,
+           offset * sizeof(float));
+    w_data += offset;
+
+    // copy wh data
```
Comment (Superjomn) [inline paddle/fluid/framework/ir/fc_cudnn_gru_fuse_pass.cc]: This long function mixes too many operations:

- Prepare weight
- construct the OpDesc

Each one is very long, please split to several shorter functions.

File: paddle/fluid/framework/ir/fc_cudnn_gru_fuse_pass.cc [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+    std::vector<int64_t> wdim = {weight_x_tensor->dims()[0] * 3 +
+                                     weight_h_tensor->dims()[1] +
+                                     gru_bias_tensor->dims()[0] * 6,
+                                 weight_h_tensor->dims()[0]};
+
+    framework::DDim dim = framework::make_ddim(wdim);
+    w_tensor->Resize(dim);
+
+    w_tensor->mutable_data<float>(platform::CPUPlace());
```
Comment (Superjomn) [inline paddle/fluid/framework/ir/fc_cudnn_gru_fuse_pass.cc:175]: merge L108 and 114

File: paddle/fluid/operators/cudnn_gru_op.cu.cc [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+
+    auto &dev_ctx = ctx.template device_context<platform::CUDADeviceContext>();
+    auto handle = dev_ctx.cudnn_handle();
+    auto *cache_var = ctx.InputVar("Cache");
+    if (!cache_var) {
+      // The RAW type cache variable wouldn't be created and broadcasted on
+      // multi-devices before the first running.
+      // use parent scope to make cache persistable
+      auto *scope = const_cast<framework::Scope *>(ctx.scope().parent());
+      auto cache_var_name = ctx.Inputs("Cache")[0];
```
Comment (Superjomn) [inline paddle/fluid/operators/cudnn_gru_op.cu.cc]: assert ctx.Inputs("Cache") is not empty first

ctx.HasInput("Cache") ?

File: paddle/fluid/operators/cudnn_gru_op.cu.cc [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+    bool is_test = ctx.Attr<bool>("is_test");
+
+    auto &dev_ctx = ctx.template device_context<platform::CUDADeviceContext>();
+    auto handle = dev_ctx.cudnn_handle();
+    auto *cache_var = ctx.InputVar("Cache");
+    if (!cache_var) {
+      // The RAW type cache variable wouldn't be created and broadcasted on
+      // multi-devices before the first running.
+      // use parent scope to make cache persistable
+      auto *scope = const_cast<framework::Scope *>(ctx.scope().parent());
```
Comment (Superjomn) [inline paddle/fluid/operators/cudnn_gru_op.cu.cc]: assert scope != null before using `scope->Var`

File: paddle/fluid/operators/cudnn_gru_op.cu.cc [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+      std::random_device rnd;
+      int seed = ctx.Attr<int>("seed");
+      if (seed == -1) {
+        seed = rnd();
+      }
+
+      auto input_w_numel = w->numel();
+      int batch_size = 1;
+      // if input not lod tensor
+      if (x->lod().size() == 0) {
```
Comment (Superjomn) [inline paddle/fluid/operators/cudnn_gru_op.cu.cc]: x->lod().empty()

File: paddle/fluid/operators/cudnn_gru_op.cu.cc [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+        batch_size = x_dim[1];
+      } else {
+        framework::LoD x_lod = x->lod();
```
Comment (Superjomn) [inline paddle/fluid/operators/cudnn_gru_op.cu.cc]: The lod related code can split to a function

File: paddle/fluid/operators/cudnn_gru_op.cu.cc [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+        }
+      }
+      cudnn_rnn_cache->init(handle, ctx.GetPlace(), max_len, batch_size,
+                            input_size, hidden_size, num_layers, dropout_prob,
+                            is_bidirec, seed, input_w_numel, true);
+    }
+
+    auto run_seq_len = x->dims()[0];
+    T *out_data = out->mutable_data<T>(ctx.GetPlace());
+    if (seq_len.size() > 0) {
```
Comment (Superjomn) [inline paddle/fluid/operators/cudnn_gru_op.cu.cc]: !seq_len.empty()

File: paddle/fluid/operators/cudnn_gru_op.cu.cc [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+      int x_padding_size = max_seq_len * x_dim[1];
+      cudaErrCheck(cudaMalloc(reinterpret_cast<void **>(&x_data_padding),
+                              x_padding_size));
+      // do cudnn for each lod, maybe need refine.
+      for (auto &len : seq_len) {
+        int mem_size = len * x_dim[1];
+        cudaErrCheck(cudaMemcpy(x_data_padding, x_data, mem_size * sizeof(T),
```
Comment (Superjomn) [inline paddle/fluid/operators/cudnn_gru_op.cu.cc]: PADDLE_ENFORCE has the check for cuda related call

No need to call the raw cuda check API

File: paddle/fluid/operators/cudnn_gru_op.cu.cc [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+#include "paddle/fluid/operators/cudnn_rnn_cache.h"
+#include "paddle/fluid/operators/math/math_function.h"
+
+namespace paddle {
+namespace operators {
+
+using LoDTensor = framework::LoDTensor;
+using Tensor = framework::Tensor;
+
+#define cudaErrCheck(stat) \
```
Comment (Superjomn) [inline paddle/fluid/operators/cudnn_gru_op.cu.cc]: use PADDLE_ENFORCE instead, https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/platform/enforce.h#L160

File: paddle/fluid/operators/cudnn_gru_op.cu.cc [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+
+    const T *x_data = x->data<T>();
+    const T *init_h_data = nullptr;
+    if (init_h->numel() != 0) {
+      init_h_data = init_h->data<T>();
+    }
+
+    const T *w_data = w->data<T>();
+
+    // T *last_h_data = last_h->mutable_data<T>(ctx.GetPlace());
```
Comment (Superjomn) [inline paddle/fluid/operators/cudnn_gru_op.cu.cc]: remove this

**Review by Superjomn — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/cudnn_gru_op.cu.cc (hunk 5 above)
Comment (zhaoyuchen2018) [reply paddle/fluid/operators/cudnn_gru_op.cu.cc]: InferShape has asserted this attribute

File: paddle/fluid/inference/api/paddle_pass_builder.h (hunk 1 above)
Comment (zhaoyuchen2018) [reply paddle/fluid/inference/api/paddle_pass_builder.h]: updated, please help to review, thanks.

**Pull request closed without merging**